datafusion/core/src/physical_plan/joins/hash_join.rs

```diff
@@ -1440,44 +1181,150 @@ fn equal_rows(
+            // combine the matched and unmatched right result together
+            append_right_indices(left_indices, right_indices, right_unmatched_indices)
+        }
+        JoinType::RightSemi => {
+            // need to remove the duplicated record in the right side
+            let right_indices = get_semi_indices(count_right_batch, &right_indices);
+            // the left_indices will not be used later for the `right semi` join
+            (left_indices, right_indices)
+        }
+        JoinType::RightAnti => {
+            // need to remove the duplicated record in the right side
+            // get the anti index for the right side
+            let right_indices = get_anti_indices(count_right_batch, &right_indices);
+            // the left_indices will not be used later for the `right anti` join
+            (left_indices, right_indices)
+        }
+        JoinType::LeftSemi | JoinType::LeftAnti => {
+            // matched or unmatched left row will be produced in the end of loop
+            // TODO: left semi can be optimized.
+            // When visit the right batch, we can output the matched left row and don't need to wait the end of loop
+            (
+                UInt64Array::from_iter_values(vec![]),
+                UInt32Array::from_iter_values(vec![]),
+            )
+        }
+    }
+}
+
+fn append_right_indices(
+    left_indices: UInt64Array,
+    right_indices: UInt32Array,
+    appended_right_indices: UInt32Array,
+) -> (UInt64Array, UInt32Array) {
+    // left_indices, right_indices and appended_right_indices must not contain the null value
+    if appended_right_indices.is_empty() {
+        (left_indices, right_indices)
     } else {
-        UInt64Array::from_iter_values(
-            (0..visited_left_side.len())
-                .filter_map(|v| (visited_left_side.get_bit(v)).then_some(v as u64)),
-        )
-    };
+        let matched_size = left_indices.len();
+        let unmatched_size = appended_right_indices.len();
+        let total_size = matched_size + unmatched_size;
+        // the new left indices: left_indices + null array
+        // the new right indices: right_indices + appended_right_indices
+        let new_left_indices = (0..total_size)
+            .map(|pos| {
+                if pos < matched_size {
+                    unsafe { Some(left_indices.value_unchecked(pos)) }
+                } else {
+                    None
+                }
+            })
+            .collect::<UInt64Array>();
+        let new_right_indices = (0..total_size)
+            .map(|pos| {
+                if pos < matched_size {
+                    unsafe { Some(right_indices.value_unchecked(pos)) }
+                } else {
+                    unsafe {
+                        Some(appended_right_indices.value_unchecked(pos - matched_size))
+                    }
+                }
+            })
+            .collect::<UInt32Array>();
+        (new_left_indices, new_right_indices)
```

Comment:
I think you might be able to do this without unsafe and more concisely:

```suggestion
        let matched_size = left_indices.len();
        let unmatched_size = appended_right_indices.len();
        // the new left indices: left_indices + null array
        // the new right indices: right_indices + appended_right_indices
        let new_left_indices = left_indices
            .iter()
            .chain(std::iter::repeat(None).take(unmatched_size))
            .collect::<UInt64Array>();
        let new_right_indices = right_indices
            .iter()
            .chain(appended_right_indices.iter())
            .collect::<UInt32Array>();
        (new_left_indices, new_right_indices)
```

Reply:
👍